src/dbg/TraceRecord.cpp

```diff
@@ -0,0 +1,445 @@
+        {
+        case TraceRecordType::TraceRecordBitExec:
+            size = 4096 / 8;
+            break;
+        case TraceRecordType::TraceRecordByteWithExecTypeAndCounter:
+            size = 4096;
+            break;
+        case TraceRecordType::TraceRecordWordWithExecTypeAndCounter:
+            size = 4096 * 2;
+            break;
+        default:
+            size = 0;
+            break;
+        }
+        if (size != 0)
+        {
+            currentPage.rawPtr = emalloc(size, "TraceRecordManager");
+            const char *p = json_string_value(json_object_get(value, "data"));
+            char *read_ptr = (char*)currentPage.rawPtr;
+            const char *c;
+            for (c = p; c < p + size * 2 && *c != '\0'; c += 2, read_ptr++)
+            {
+                if (c[1] >= '0' && c[1] <= '9')
+                    *read_ptr = c[1] - '0';
+                else if (c[1] >= 'A' && c[0] <= 'F')
+                    *read_ptr = c[1] - 'A' + 10;
+                else
+                    break;
+                if (c[0] >= '0' && c[0] <= '9')
+                    *read_ptr |= (c[0] - '0') << 4;
+                else if (c[0] >= 'A' && c[0] <= 'F')
+                    *read_ptr |= (c[0] - 'A' + 10) << 4;
+                else
+                    break;
+            }
+            if (c != p + size * 2)
+            {
+                efree(currentPage.rawPtr, "TraceRecordManager");
+            }
+            else
+            {
+                const char* moduleName = json_string_value(json_object_get(value, "module"));
+                duint key;
+                dprintf("%s\n", moduleName);
+                if (*moduleName)
+                {
+                    currentPage.moduleIndex = getModuleIndex(std::string(moduleName));
+                    key = currentPage.rva + ModHashFromName(moduleName);
+                }
+                else
+                {
+                    currentPage.moduleIndex = ~0;
+                    key = currentPage.rva;
+                }
+                TraceRecord.insert(std::make_pair(key, currentPage));
+            }
+        }
+    }
+}
+
+unsigned int TraceRecordManager::getModuleIndex(std::string moduleName)
+{
+    auto iterator = std::find(ModuleNames.begin(), ModuleNames.end(), moduleName);
+    if (iterator != ModuleNames.end())
+        return iterator - ModuleNames.begin();
+    else
+    {
+        ModuleNames.push_back(moduleName);
+        return ModuleNames.size() - 1;
+    }
+}
+
+void _dbg_dbgtraceexecute(duint CIP)
+{
+    if(TraceRecord.getTraceRecordType(CIP) != TraceRecordManager::TraceRecordType::TraceRecordNone)
+    {
+        Capstone disassembler;
+        unsigned char buffer[16];
+        duint size;
+        if(MemRead(CIP, buffer, 16))
+        {
+            TraceRecord.increaseInstructionCounter();
+            disassembler.Disassemble(CIP, buffer);
+            size = disassembler.Success() ? disassembler.Size() : 1;
```

Comment:
I don't quite understand the close to page condition. If you use it
correctly you copy min(16, sizeLeftInPage) to the 16 byte buffer which you
pass to the disasm function. Since disasm length it never bigger than 16
there should never be any problems.
On Wed, 25 May 2016 at 18:55, Torusrxxx <notifications@github.com> wrote:

> In src/dbg/TraceRecord.cpp
> <https://github.com/x64dbg/x64dbg/pull/665#discussion_r64608641>:
>
> > +        return ModuleNames.size() - 1;
> > +    }
> > +}
> > +
> > +void _dbg_dbgtraceexecute(duint CIP)
> > +{
> > +    if(TraceRecord.getTraceRecordType(CIP) != TraceRecordManager::TraceRecordType::TraceRecordNone)
> > +    {
> > +        Capstone disassembler;
> > +        unsigned char buffer[16];
> > +        duint size;
> > +        if(MemRead(CIP, buffer, 16))
> > +        {
> > +            TraceRecord.increaseInstructionCounter();
> > +            disassembler.Disassemble(CIP, buffer);
> > +            size = disassembler.Success() ? disassembler.Size() : 1;
>
> So the current code reads the current instruction, handles the corner case
> correctly, and without using that buffer it just calls the same routine
> which is not aware of close-to-page-end condition. A boring code
> duplication and waste.
>
> —
> You are receiving this because you modified the open/close state.
>
>
> Reply to this email directly or view it on GitHub
> <https://github.com/x64dbg/x64dbg/pull/665/files/2359e6a6edf18bc30358eab576cce2b46d646f0c#r64608641>
>